src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -410,16 +410,19 @@ double SolverInterfaceImpl::advance(
   // Current time
   double time = _couplingScheme->getTime();
 
-  if (_couplingScheme->willDataBeExchanged(0.0)) {
-    performDataActions({action::Action::WRITE_MAPPING_PRIOR}, time);
-    mapWrittenData();
-    performDataActions({action::Action::WRITE_MAPPING_POST}, time);
-  }
+  // always perform mapping, because we will store all mapped data, if subcycling is used.
+  performDataActions({action::Action::WRITE_MAPPING_PRIOR}, time);
+  mapWrittenData();
+  performDataActions({action::Action::WRITE_MAPPING_POST}, time);
 
   PRECICE_DEBUG("Advance coupling scheme");
   _couplingScheme->advance();
 
-  if (_couplingScheme->isTimeWindowComplete()) {
+  if (_couplingScheme->isTimeWindowComplete() && _couplingScheme->moveWindowBeforeMapping()) { /*
+     * move to next window before mapping, if second in serial coupling.
+     * Reason:
+     * Window is complete, but first participant has already performed an iteration in the new window and the data just received is already from the new window. So we want to use it, but before doing this we have to use the data from the old window to initialize the waveform at 0.0.
+     */
```

Comment:
Moving the `moveToNextWindow()` call inside the coupling scheme is complicated, because we have to be able to access the `ReadDataContext` where the waveform lives in order to tell it to move to the next window. Alternatively, we could give the coupling scheme access to the `ReadDataContext`, but this would lead to more interaction between the different modules than necessary and lead to some degree of spaghettification. I would rather keep this solution, because here the `SolverInterfaceImpl` clearly stays in control and there is no direct interaction of the coupling scheme and the waveform.

Reply:
> So we want to use it, but before doing this we have to use the data from the old window to initialize the waveform at 0.0.

Isn't there a solution where we do this at the end of the previous time window. So in the call below?
Let's draw a picture together :blush: 

Anyway, this surely will give problems with `CompostionalCouplingScheme`. We need a different solution, sorry :confused:

Reply:
Then we should also try to come up with an integration test for the compositional coupling scheme + time interpolation. This makes it easier to agree on what we want and it also makes sure that the feature will not break in the future without noticing it.

Moving this to the previous time window: Not sure. It's definitely complicated and I don't see a solution with digging deep into the problem.

Reply:
I just did two brief experiments:

1) Completely removing the block `if (_couplingScheme->isTimeWindowComplete() && _couplingScheme->moveWindowBeforeMapping()) { ... }`
2) Moving the block `if (_couplingScheme->isTimeWindowComplete() && _couplingScheme->moveWindowBeforeMapping()) { ... }` before `_couplingScheme->advance()`

In both cases we get failing tests for `mpirun -np 4 ./testprecice -t Integration/Serial/Time/Implicit/SerialCoupling`. I already expected this and it means:

* The feature is actually needed
* The convergence communication that happens inside advance is needed for the correct evaluation of `_couplingScheme->moveWindowBeforeMapping()`
* It's unfortunately not that simple ;)

@uekerman and @fsimonis you are currently also working on splitting up advance, correct? Might this be helpful in this context? I think generally there should not be a big problem here and all the information that we need about convergence is available when we need it. It's still a bit tricky with all the communication going on.